### PR TITLE
Merge in most of realm-js-private

### PR DESCRIPTION
**Pull request by tgoyne**
An overview of these changes:

The JS bindings and tests for the adapter and global notifier are now here, along with a few assorted functions which didn't realy have any reason to be in realm-js-private to begin with. The adapter and global notifier themselves are now in object store.

All of the stuff around building with sync enabled on only macOS or non-linux (depending on where the check happened to be made) is gone as that was a relic of Developer Edition.

The real build system for node is now here. Previous there were two seperate gyp build systems (one here and one in realm-js-private) and we ran the tests against the one here but built releases with the other one. Unsurprisingly this resulted in the tests failing to catch build-system related issues and was a giant headache.

All of the release packaging stuff in the Jenkinsfile is now here and tagging releases here should produce actual release artifacts on s3 (but that's untested).

A bunch of assorted test updates:
1. Upgraded from jasmine 2 to jasmine 3. This adds native async test support so that we can drop jasmine-co.
2. ROS is now run via a custom launcher script like cocoa and java did already to let us customize ROS setup more. In particular, this lets us actually enable and test JWT auth.
3. Tests now put all their files in various temporary directories rather than dumping a bunch of stuff in the realm-js repo.
4. Integration tests for node and electron now test the actual node-pre-gyp binary that'll be uploaded to S3 rather than building it from source. Building from source for local testing is still supported, and the react-native integration tests always build from source (since we don't package prebuilt binaries for that).
5. Fixed some flakey and broken tests. Turns out a lot of the tests relied on specific timing to work without anything enforcing that ordering.
6. A bunch of Dockerfile updates: previously realm-js used an image based on Ubuntu, while realm-js-private used a CentOS base image, and neither supported everything our test/packaging scripts needed. They've now been combined into a single centos-based image. The Android image remains separate and mostly unchanged. All the scripting around managing docker images has been replaced with the versions from realm-js-private which did the same things as the scripts here but in different ways.

realm-js-private still has a few assorted scripts not merged here (such as the token generation for ROS) that should maybe get shunted off somewhere else so that that repo can be archived.

### CODE REVIEW

**Comment by tgoyne** [comment]
I think the RN Android test failing is just an issue with the test device (it was working earlier, but now is hanging forever in `adb wait-for-device`).

File: tests/js/support/test-object-server.js [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            const objects = realm.objects('RealmFile').filtered('path = $0', serverPath);
+            if (objects.length != 1) {
+                throw new Error(`Realm ${serverPath} not found. Existing Realms: ${realm.objects('RealmFile').map(r => r.path)}`);
+            }
+            realm.delete(objects);
+        });
+
+        // Tell the sync worker to delete the Realm
+        const request = http.request({
+            host: 'localhost',
```
Comment (kneth) [inline tests/js/support/test-object-server.js]: `127.0.0.1` maybe?

File: scripts/nvm-wrapper.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  exit 1
+}
+
+if [[ $# -lt 2 ]]; then
+  usage
+  exit 1
+fi
+
+if [ "$(uname)" = 'Darwin' ]; then
+  source "$(brew --prefix nvm)/nvm.sh"
```
[the remark this answers is not in the document]
Comment (tgoyne) [reply scripts/nvm-wrapper.sh]: This is just a helper script for the Jenkinsfile and isn't used by anything that you can actually run locally, so it only needs to support our Jenkins worker environment.

**Comment by ironage** [comment]
If I'm reading the CI output correctly, there is a problem in the React Native Android Integration tests:
```
integration tests / React Native on Android / "before each" hook for "is a function" – "before each" hook for "is a function"
<1s
Error
remove_dir_recursive() failed: No such file or directory Path: /data/data/io.realm.tests.reactnative/files/realm-object-server
Stacktrace
clearTestState@[native code]
http://localhost:8081/index.delta?platform=android&dev=true&minify=false:105000:25
```
The other failure is one of those jnlp transient errors that will probably go away on a rerun.
Is there anything else blocking this work? It is really valuable and would really nice to merge this soon.

**Comment by tgoyne** [comment]
It was waiting on https://github.com/realm/realm-js-private/pull/528, but now that's incorporated into this. Now I just need to verify that it *can* pass (even if it doesn't every time...) as it now actually runs all the tests as part of making a release.

**Comment by tgoyne** [comment]
Well it passed so I'm going to mash that merge button. I hope this doesn't break anything.

**Pull request merged**